ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -16,45 +18,53 @@ import           Control.Monad
 import           Control.Monad.Class.MonadFork
 import           Control.Monad.Class.MonadSay
 import           Control.Monad.Class.MonadSTM
-import           Control.Monad.Class.MonadTimer
 import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Class.MonadTimer
 import           Data.Bits
 import qualified Data.ByteString.Lazy as BL
 import           Data.Int
 import           Data.Word
+import           GHC.Stack
 import           Network.Socket hiding (recv, recvFrom, send, sendTo)
 import qualified Network.Socket.ByteString.Lazy as Socket (recv, sendAll)
 
+import           Network.TypedProtocol.Driver
 import           Ouroboros.Network.Chain (Chain, ChainUpdate, Point)
 import qualified Ouroboros.Network.Chain as Chain
 import qualified Ouroboros.Network.ChainProducerState as CPS
+import           Ouroboros.Network.Channel
 import qualified Ouroboros.Network.Mux as Mx
 import           Ouroboros.Network.Protocol.ChainSync.Client
 import           Ouroboros.Network.Protocol.ChainSync.Codec
 import           Ouroboros.Network.Protocol.ChainSync.Examples
 import           Ouroboros.Network.Protocol.ChainSync.Server
 import           Ouroboros.Network.Serialise
-import           Ouroboros.Network.Channel
-import           Network.TypedProtocol.Driver
 
 import           Text.Printf
 
 newtype SocketCtx = SocketCtx { scSocket :: Socket }
 
-setupMux :: (DemoProtocols -> Mx.MiniProtocolDescription DemoProtocols IO)
-         -> SocketCtx -> IO ()
-setupMux mpds ctx = do
+setupMux :: (Mx.ProtocolEnum ptcl, Ord ptcl, Enum ptcl, Bounded ptcl) => (ptcl -> Mx.MiniProtocolDescription ptcl IO)
+         -> SocketCtx
+         -> Maybe (TQueue IO (Maybe SomeException))
```

Comment:
Maybe instead of `Maybe (TBQueue IO (Maybe SomeException))` it would be more convenient to have a `Maybe SomeException -> IO ()` callback.

Reply:
I like this idea, will make the change.